pkg/chaosctl/debug/stresschaos/stresschaos.go

```diff
@@ -0,0 +1,167 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stresschaos
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"code.cloudfoundry.org/bytefmt"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	cm "github.com/chaos-mesh/chaos-mesh/pkg/chaosctl/common"
+)
+
+// Debug get chaos debug information
+func Debug(ctx context.Context, chaos runtime.Object, c *cm.ClientSet, result *cm.ChaosResult) error {
+	stressChaos, ok := chaos.(*v1alpha1.StressChaos)
+	if !ok {
+		return fmt.Errorf("chaos is not stresschaos")
+	}
+	chaosStatus := stressChaos.Status.ChaosStatus
+	chaosSelector := stressChaos.Spec.GetSelector()
+
+	pods, daemons, err := cm.GetPods(ctx, chaosStatus, chaosSelector, c.CtrlCli)
+	if err != nil {
+		return err
+	}
+
+	if err := cm.CheckFailedMessage(ctx, chaosStatus.FailedMessage, daemons, c); err != nil {
+		return err
+	}
+
+	for i := range pods {
+		podName := pods[i].Name
+		podResult := cm.PodResult{Name: podName}
+		err := debugEachPod(ctx, pods[i], daemons[i], stressChaos, c, &podResult)
+		result.Pods = append(result.Pods, podResult)
+		if err != nil {
+			return fmt.Errorf("for %s: %s", podName, err.Error())
+		}
+	}
+	return nil
+}
+
+func debugEachPod(ctx context.Context, pod v1.Pod, daemon v1.Pod, chaos *v1alpha1.StressChaos, c *cm.ClientSet, result *cm.PodResult) error {
+	// cpu or memory chaos
+	isCPU := true
+	if cpuSpec := chaos.Spec.Stressors.CPUStressor; cpuSpec == nil {
+		isCPU = false
+	}
+
+	// get process path
+	cmd := fmt.Sprintf("cat /proc/cgroups")
+	out, err := cm.Exec(ctx, pod, daemon, cmd, c.KubeCli)
+	if err != nil {
+		return fmt.Errorf("run command '%s' failed with: %s", cmd, err.Error())
+	}
+	var cpuMountType string
+	if regexp.MustCompile("(cpu,cpuacct)").MatchString(string(out)) {
+		cpuMountType = "cpu,cpuacct"
+	} else {
+		cpuMountType = "cpu"
+	}
+
+	cmd = fmt.Sprintf("ps")
+	out, err = cm.Exec(ctx, pod, daemon, cmd, c.KubeCli)
+	if err != nil {
+		return fmt.Errorf("run command '%s' failed with: %s", cmd, err.Error())
+	}
+	stressngLine := regexp.MustCompile("(.*)(stress-ng)").FindStringSubmatch(string(out))
+	if len(stressngLine) == 0 {
+		return fmt.Errorf("Could not find stress-ng, StressChaos failed")
+	}
+	stressngPid := strings.Fields(stressngLine[0])[0]
+
+	cmd = fmt.Sprintf("cat /proc/1/cgroup")
+	out, err = cm.Exec(ctx, pod, daemon, cmd, c.KubeCli)
+	if err != nil {
+		return fmt.Errorf("run command '%s' failed with: %s", cmd, err.Error())
+	}
+	result.Items = append(result.Items, cm.ItemResult{Name: "/proc/1/cgroup", Value: string(out)})
```

Comment:
Hi @STRRL could you give me some reference about the time when PID 1 is not our target. Sorry I couldn't find those myself 😿 

For this problem, print `/proc/cgroups` could work well, but for the later one, which needs to list `/proc/$PID/fd`, is there any method that we could get to fetch the `init` process when the pid is not 1.

Reply:
> Hi @STRRL could you give me some reference about the time when PID 1 is not our target.

You could take a look at this repo: [hello-kubernetes](https://github.com/paulbouwer/hello-kubernetes), it's a node web application. `npm` is the PID 1 process, and `node` is a child process.

>For this problem, print /proc/cgroups could work well, but for the later one, which needs to list /proc/$PID/fd, is there any method that we could get to fetch the init process when the pid is not 1.

How about printing fd information for each process?

Reply:
Thank you @STRRL for the example! Learned a lot!
> How about printing fd information for each process?

I'm not so sure, it could be a mess when the target pod is complicated. But if maintainers could easily filter out what they need, I'm ok with this way.
cc @YangKeao what do you think

Reply:
Sorry I made a mistake here. We need to compare cgroup of "the application" with `stress-ng`, as shown in [the original post](https://github.com/chaos-mesh/chaos-mesh/issues/755#issuecomment-667006237). So we still need to fetch the target pid.

Reply:
> I'm not so sure, it could be a mess when the target pod is complicated.

"Mess" is not a big issue, it just a thing that needs improvement with UX. 

Imagination: When you profiling a production environment:

- your tools dump a lot of information, you will finally find the reason by your own eyeball.
- your tools print not enough information, then you request your "System Administrator" for SSH permission for next profiling. ¯\\_(ツ)_/¯

Then talking about the UX, `grep` is the CLI tool that we always use. So a "grep-friendly" tool is our favorite. For example:

```text
# bad 🙁
[StressChaos for pod-xxx-xxxxx]
    cgroups:
        cpu: xxx
        cpuacct: xxx
        memory: xxx

# good 😆
StressChaos for pod-xxx-xxxxx cgroups cpu: xxx
StressChaos for pod-xxx-xxxxx cgroups cpuacct: xxx
StressChaos for pod-xxx-xxxxx cgroups memory: xxx
```

If you interest in it, we could create an issue and make discussions about this. And We could improve UX as another PR.

> Sorry I made a mistake here. We need to compare cgroup of "the application" with stress-ng, as shown in the original post. So we still need to fetch the target pid.

I still suggest print the all information. 🤪

Reply:
Thank you I never thought about the way "grep-friendly" before! It's not hard to change I'll commit about this later.

> I still suggest print the all information.

What if we set a flag for this, so users could explicitly set their application PID. Then the problem would turn to by default do we print all or print 1.

Reply:
> Considering this tool is mainly for users to provide debug information, so IDE or just browser is what we use to filter out information. In this case I think a better hierarchy might be more intuitive?

It's not so emergency, I suggest make this in another PR. : )

As you already refactored the codes: from " just printing it" to "using structs to describe information then "rendering" it to stdout". So you can easily change the way of rendering.

Reply:
Sure thx I'll note it down for later.

Reply:
Hi @STRRL print all info fixed PTAL